### ISSUE TITLE
Raspberry Pi - incorrect categorization of input/output device, and unknown default device

### ISSUE THREAD

**Issue by andrewrk** [andrewrk/libsoundio#44]
@czxJAVA2009 writes:

> I can't build the libsoundio.After I make it.following error:
```
[ 2%] Building CXX object CMakeFiles/libsoundio_shared.dir/src/soundio.cpp.o
In file included from /home/orangepi/libsoundio/src/alsa.hpp:13:0,
from /home/orangepi/libsoundio/src/soundio.hpp:23,
from /home/orangepi/libsoundio/src/soundio.cpp:8:
/home/orangepi/libsoundio/src/atomics.hpp:27:2: error: #error "require atomic_bool to be lock free"
#error "require atomic_bool to be lock free"
^
In file included from /home/orangepi/libsoundio/src/soundio.hpp:23:0,
from /home/orangepi/libsoundio/src/soundio.cpp:8:
/home/orangepi/libsoundio/src/alsa.hpp:48:5: error: ‘snd_pcm_chmap_t’ does not name a type
snd_pcm_chmap_t chmap;
^
/home/orangepi/libsoundio/src/alsa.hpp:68:5: error: ‘snd_pcm_chmap_t’ does not name a type
snd_pcm_chmap_t *chmap;
^
cc1plus: error: unrecognized command line option "-Wno-c99-extensions" [-Werror]
cc1plus: all warnings being treated as errors
CMakeFiles/libsoundio_shared.dir/build.make:62: recipe for target 'CMakeFiles/libsoundio_shared.dir/src/soundio.cpp.o' failed
make[2]: ** [CMakeFiles/libsoundio_shared.dir/src/soundio.cpp.o] Error 1
CMakeFiles/Makefile2:205: recipe for target 'CMakeFiles/libsoundio_shared.dir/all' failed
make[1]: *** [CMakeFiles/libsoundio_shared.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by diogocp** [comment]
Old alsa? #30

**Comment by andrewrk** [comment]
I'm concerned about the error about atomic_bool not being lock free. I have a pi and I'm installing fresh raspbian on it. It'll be a good test to make sure libsoundio works correctly on that hardware.

**Comment by andrewrk** [comment]
Indeed, the gcc version of in current raspbian is too old and does not have lock-free atomics. However, the clang version is able to compile libsoundio. You can do it like this:

```
mkdir build
cd build
CC=clang CXX=clang++ ..
make
```

I ran into another issue, when building ./unit_tests:

```
/usr/bin/ld: cannot find /usr/lib/llvm-3.5/bin/../lib/clang/3.5.0/lib/linux/libclang_rt.profile-arm.a: No such file or directory
```

Looks like a bug in the raspbian clang package. Workaround is to compile without test coverage report:

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8959a9e..f236d8a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -205,8 +205,6 @@ set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -Werror -pedantic")
 
 set(LIB_CFLAGS "-std=c++11 -fno-exceptions -fno-rtti -fvisibility=hidden -Wall -Werror=strict-prototypes -Werror=old-style-definition -Werror=missing-prototypes -Wno-c99-extensions")
 set(EXAMPLE_CFLAGS "-std=c99 -Wall")
-set(TEST_CFLAGS "${LIB_CFLAGS} -fprofile-arcs -ftest-coverage")
-set(TEST_LDFLAGS "-fprofile-arcs -ftest-coverage")
 
 configure_file(
     "${CMAKE_SOURCE_DIR}/src/config.h.in"
@@ -281,8 +279,7 @@ if(BUILD_TESTS)
     target_link_libraries(unit_tests LINK_PUBLIC ${LIBSOUNDIO_LIBS})
     set_target_properties(unit_tests PROPERTIES
         LINKER_LANGUAGE C
-        COMPILE_FLAGS ${TEST_CFLAGS}
-        LINK_FLAGS ${TEST_LDFLAGS}
+        COMPILE_FLAGS ${LIB_CFLAGS}
     )
 
     add_executable(latency "${CMAKE_SOURCE_DIR}/test/latency.cpp" ${LIBSOUNDIO_SOURCES})
```

Now it looks like I have some work to do, because ./unit_tests fails with:

```
testing os_get_time...OK
testing create output stream...ALSA lib pcm_dsnoop.c:557:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:557:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
unit_tests: /home/pi/dev/libsoundio/test/unit_tests.cpp:37: void test_create_outstream(): Assertion `default_out_device_index >= 0' failed.
Aborted
```

./sio_sine gives:

```
Backend: ALSA
ALSA lib pcm_dsnoop.c:557:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:557:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
Output device not found
```

Looks like what's going on here is two problems:

0. The default device was not detected correctly.
0. An input device was mistaken for an output device.

Sure enough:

```
pi@raspberrypi ~/dev/libsoundio/build $ ./sio_list_devices 
ALSA lib pcm_dsnoop.c:557:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
ALSA lib pcm_dsnoop.c:557:(snd_pcm_dsnoop_open) The dsnoop plugin supports only capture stream
--------Input Devices--------


--------Output Devices--------

bcm2835 ALSA, bcm2835 ALSA: Direct sample mixing device
  id: dmix:CARD=ALSA,DEV=0
  channel layouts:
    Stereo
  sample rates:
    48000 - 48000
  current sample rate: 48000
  formats: signed 16-bit LE
  min software latency: 0.04266667 sec
  max software latency: 0.34133333 sec

bcm2835 ALSA, bcm2835 IEC958/HDMI: Direct sample mixing device
  id: dmix:CARD=ALSA,DEV=1
  channel layouts:
    Stereo
  sample rates:
    48000 - 48000
  current sample rate: 48000
  formats: signed 16-bit LE
  min software latency: 0.04266667 sec
  max software latency: 0.34133333 sec

bcm2835 ALSA, bcm2835 ALSA: Direct sample snooping device
  id: dsnoop:CARD=ALSA,DEV=0
  probe error: unable to open device

bcm2835 ALSA, bcm2835 IEC958/HDMI: Direct sample snooping device
  id: dsnoop:CARD=ALSA,DEV=1
  probe error: unable to open device

bcm2835 ALSA, bcm2835 ALSA: Direct hardware device without any conversions
  id: hw:CARD=ALSA,DEV=0
  channel layouts:
    Mono
    Stereo
  sample rates:
    8000 - 48000
  formats: signed 16-bit LE
  min software latency: 0.00533333 sec
  max software latency: 0.68266667 sec

bcm2835 ALSA, bcm2835 IEC958/HDMI: Direct hardware device without any conversions
  id: hw:CARD=ALSA,DEV=1
  channel layouts:
    Stereo
  sample rates:
    44100 - 48000
  formats: signed 16-bit LE
  min software latency: 0.00533333 sec
  max software latency: 0.34133333 sec

bcm2835 ALSA, bcm2835 ALSA: Hardware device with all software conversions
  id: plughw:CARD=ALSA,DEV=0
  channel layouts:
    Mono
    Stereo
    2.1
    3.0
    3.0 (back)
    3.1
    4.0
    Quad
    Quad (side)
    4.1
    5.0 (back)
    5.0 (side)
    5.1
    5.1 (back)
    6.0 (side)
    6.0 (front)
    Hexagonal
    6.1
    6.1 (back)
    6.1 (front)
    7.0
    7.0 (front)
    7.1
    7.1 (wide)
    7.1 (wide) (back)
    Octagonal
  sample rates:
    8000 - 48000
  formats: signed 8-bit, unsigned 8-bit, signed 16-bit LE, signed 16-bit BE, unsigned 16-bit LE, unsigned 16-bit LE, signed 24-bit LE, signed 24-bit BE, unsigned 24-bit LE, unsigned 24-bit BE, signed 32-bit LE, signed 32-bit BE, unsigned 32-bit LE, unsigned 32-bit BE, float 32-bit LE, float 32-bit BE, float 64-bit LE, float 64-bit BE
  min software latency: 0.00533333 sec
  max software latency: 0.68266667 sec

bcm2835 ALSA, bcm2835 IEC958/HDMI: Hardware device with all software conversions
  id: plughw:CARD=ALSA,DEV=1
  channel layouts:
    Mono
    Stereo
    2.1
    3.0
    3.0 (back)
    3.1
    4.0
    Quad
    Quad (side)
    4.1
    5.0 (back)
    5.0 (side)
    5.1
    5.1 (back)
    6.0 (side)
    6.0 (front)
    Hexagonal
    6.1
    6.1 (back)
    6.1 (front)
    7.0
    7.0 (front)
    7.1
    7.1 (wide)
    7.1 (wide) (back)
    Octagonal
  sample rates:
    44100 - 48000
  formats: signed 8-bit, unsigned 8-bit, signed 16-bit LE, signed 16-bit BE, unsigned 16-bit LE, unsigned 16-bit LE, signed 24-bit LE, signed 24-bit BE, unsigned 24-bit LE, unsigned 24-bit BE, signed 32-bit LE, signed 32-bit BE, unsigned 32-bit LE, unsigned 32-bit BE, float 32-bit LE, float 32-bit BE, float 64-bit LE, float 64-bit BE
  min software latency: 0.00533333 sec
  max software latency: 0.34133333 sec

bcm2835 ALSA bcm2835 ALSA (raw)
  id: hw:0,0
  channel layouts:
    Mono
    Stereo
  sample rates:
    8000 - 48000
  formats: signed 16-bit LE
  min software latency: 0.00533333 sec
  max software latency: 0.68266667 sec

bcm2835 ALSA bcm2835 IEC958/HDMI (raw)
  id: hw:0,1
  channel layouts:
    Stereo
  sample rates:
    44100 - 48000
  formats: signed 16-bit LE
  min software latency: 0.00533333 sec
  max software latency: 0.34133333 sec


10 devices found
```

dsnoop should be listed in input devices, not output devices. And the default should probably be `dmix:CARD=ALSA,DEV=0`, not `dsnoop:CARD=ALSA,DEV=0`.

**Comment by andrewrk** [comment]
Now that #45 is complete, libsoundio builds out-of-the-box for a Jessie Raspbian image. Note that https://github.com/andrewrk/libsoundio/issues/7#issuecomment-137569996 is still relevant for Debian Jessie. That is, if you compile with JACK support, you'll have to make a one-line patch to either libsoundio or JACK to work around a JACK bug.

So, now there are only two items remaining to fix for the pi:

 * [ ] `bcm2835 ALSA, bcm2835 ALSA: Direct sample snooping device` is listed as an output device rather than an input device.
 * [ ] The default output device is either wrong or missing.

**Comment by andrewrk** [comment]
The driver for direct sample snooping device has a bug. It's reporting itself as an output device rather than an input device. I added a workaround that detects this driver and forces it to be output only.

For the default output device, I disabled the code that forcibly ignores sysdefault. This makes listing devices output less clean on systems that have both sysdefault and default, but makes the default work on systems that have only sysdefault, which apparently the Raspberry Pi is one.

As for why I'm getting high CPU percentage and underruns, that remains to be discovered.
